src/server/gc_worker/rawkv_compaction_filter.rs

```diff
@@ -0,0 +1,514 @@
+        self.mvcc_deletions.push(key);
+    }
+
+    // TODO some refactor to avoid duplicated codes.
+    fn switch_key_metrics(&mut self) {
+        if self.versions != 0 {
+            self.versions_hist.observe(self.versions as f64);
+            self.total_versions += self.versions;
+            self.versions = 0;
+        }
+        if self.filtered != 0 {
+            self.filtered_hist.observe(self.filtered as f64);
+            self.total_filtered += self.filtered;
+            self.filtered = 0;
+        }
+    }
+
+    fn flush_metrics(&self) {
+        GC_COMPACTION_FILTERED.inc_by(self.total_filtered as u64);
+        GC_COMPACTION_FILTER_ORPHAN_VERSIONS
+            .with_label_values(&["generated"])
+            .inc_by(self.orphan_versions as u64);
+        if let Some((versions, filtered)) = STATS.with(|stats| {
+            stats.versions.update(|x| x + self.total_versions);
+            stats.filtered.update(|x| x + self.total_filtered);
+            if stats.need_report() {
+                return Some(stats.prepare_report());
+            }
+            None
+        }) {
+            if filtered > 0 {
+                info!("RawKV Compaction filter reports"; "total" => versions, "filtered" => filtered);
+            }
+        }
+    }
+}
+
+pub fn split_ts(key: &[u8]) -> Result<(&[u8], u64), String> {
+    match Key::split_on_ts_for(key) {
+        Ok((key, ts)) => Ok((key, ts.into_inner())),
+        Err(_) => Err(format!(
+            "invalid write cf key: {}",
+            log_wrappers::Value(key)
+        )),
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+
+    use super::*;
+    use api_version::RawValue;
+    use kvproto::kvrpcpb::ApiVersion;
+    use std::thread;
+
+    use crate::config::DbConfig;
+    use crate::server::gc_worker::TestGCRunner;
+    use crate::storage::kv::TestEngineBuilder;
+    use engine_traits::{Peekable, SyncMutable, CF_DEFAULT};
+    use std::time::Duration;
+    use txn_types::TimeStamp;
+
+    pub fn make_key(key: &[u8], ts: u64) -> Vec<u8> {
+        let key1 = Key::from_raw(key)
+            .append_ts(TimeStamp::new(ts))
+            .as_encoded()
+            .to_vec();
+        let res = keys::data_key(key1.as_slice());
+        res
+    }
+
+    #[test]
+    fn test_raw_compaction_filter() {
+        let mut cfg = DbConfig::default();
+        cfg.writecf.disable_auto_compactions = true;
+        cfg.writecf.dynamic_level_bytes = false;
+
+        let engine = TestEngineBuilder::new()
+            .api_version(ApiVersion::V2)
+            .build_with_cfg(&cfg)
+            .unwrap();
+        let raw_engine = engine.get_rocksdb();
+        let mut gc_runner = TestGCRunner::new(0);
+
+        let value1 = RawValue {
+            user_value: vec![0; 10],
+            expire_ts: Some(TimeStamp::max().physical()),
+            is_delete: false,
+        };
+
+        let user_key = b"r\0aaaaaaaaaaa";
+
+        raw_engine
```

Comment:
Suggest to write a test cases vector & iterate it to put entries.

Reply:
done